File: backend/src/openarchiefbeheer/destruction/api/permissions.py
from django.utils.translation import gettext_lazy as _

from rest_framework import permissions

from ..constants import ListStatus


class CanStartDestructionPermission(permissions.BasePermission):
    message = _("You are not allowed to create a destruction list.")

    def has_permission(self, request, view):
        return request.user.has_perm("accounts.can_start_destruction")


class CanReviewPermission(permissions.BasePermission):
    message = _("You are not allowed to review a destruction list.")

    def has_permission(self, request, view):
        return request.user.has_perm(
            "accounts.can_review_destruction"
        ) or request.user.has_perm("accounts.can_review_final_list")

    def has_object_permission(self, request, view, destruction_list):
        user = request.user

        # User is not assigned
        if destruction_list.assignee != user:
            return False

        # User is not permitted based on role + status
        if (
            (
                destruction_list.status == ListStatus.ready_to_review
                and not user.has_perm("accounts.can_review_destruction")
            )
            or (
                destruction_list.status == ListStatus.ready_for_archivist
                and not user.has_perm("accounts.can_review_final_list")
            )
            or destruction_list.status
            not in [ListStatus.ready_to_review, ListStatus.ready_for_archivist]
        ):
            return False

        return True


class CanCoReviewPermission(permissions.BasePermission):
    message = _("You are not allowed to co-review a destruction list.")

    def has_permission(self, request, view):
        return request.user.has_perm("accounts.can_co_review_destruction")

    def has_object_permission(self, request, view, destruction_list):
        user_assignees = destruction_list.assignees.values_list("user__pk", flat=True)
        user = request.user

        # User is not assigned
        if user.pk not in user_assignees:
            return False

        # User is not permitted based on role + status
        if (
            destruction_list.status == ListStatus.ready_to_review
            and not user.has_perm("accounts.can_co_review_destruction")
        ) or destruction_list.status != ListStatus.ready_to_review:
            return False
        return True


class CanUpdateDestructionList(permissions.BasePermission):
    message = _(
        "You are either not allowed to update this destruction list or "
        "the destruction list can currently not be updated."
    )

    def has_permission(self, request, view):
        return request.user.has_perm("accounts.can_start_destruction")

    def has_object_permission(self, request, view, destruction_list):
        return destruction_list.status == ListStatus.new


class CanMarkListAsFinal(permissions.BasePermission):
    message = _(
        "You are either not allowed to mark this destruction list as final or "
        "the destruction list can currently not be updated."
    )

    def has_permission(self, request, view):
        return request.user.has_perm("accounts.can_start_destruction")

    def has_object_permission(self, request, view, destruction_list):
        return destruction_list.status == ListStatus.internally_reviewed


class CanQueueDestruction(permissions.BasePermission):
    message = _(
        "You are either not allowed to queue the deletion of this destruction list or "
        "the destruction list can currently not be deleted."
    )

    def has_permission(self, request, view):
        return request.user.has_perm("accounts.can_start_destruction")

    def has_object_permission(self, request, view, destruction_list):
        return destruction_list.status == ListStatus.ready_to_delete


class CanDeleteList(permissions.BasePermission):
    message = _(
        "You are either not allowed to delete this destruction list or "
        "the destruction list does not have the status '%(status)s'."
    ) % {"status": ListStatus.new}

    def has_permission(self, request, view):
        return request.user.has_perm("accounts.can_start_destruction")

    def has_object_permission(self, request, view, destruction_list):
        return destruction_list.status == ListStatus.new


class CanReassignDestructionList(permissions.BasePermission):
    message = _("You are not allowed to reassign the destruction list.")

    def has_permission(self, request, view):
        return request.user.has_perm("accounts.can_start_destruction")

    def has_object_permission(self, request, view, destruction_list):
        return destruction_list.status in [
            ListStatus.new,
            ListStatus.ready_to_review,
        ]


class CanMarkAsReadyToReview(permissions.BasePermission):
    message = _("You are not allowed to mark this destruction list as ready to review.")

    def has_permission(self, request, view):
        return request.user.has_perm("accounts.can_start_destruction")

    def has_object_permission(self, request, view, destruction_list):
        return destruction_list.status == ListStatus.new


class CanAbortDestruction(permissions.BasePermission):
    message = _("You are not allowed to stop the planned destruction of the list.")

    def has_permission(self, request, view):
        return request.user.has_perm("accounts.can_start_destruction")

    def has_object_permission(self, request, view, destruction_list):
        return destruction_list.status != ListStatus.new


class CanUpdateCoReviewers(permissions.BasePermission):
    message = _("You are not allowed to update the co-reviewers.")

    def has_permission(self, request, view):
        return request.user.has_perm(
            "accounts.can_review_destruction"
        ) or request.user.has_perm("accounts.can_start_destruction")

    def has_object_permission(self, request, view, destruction_list):
        if request.user.has_perm("accounts.can_review_destruction"):
            return destruction_list.status == ListStatus.ready_to_review

        if request.user.has_perm("accounts.can_start_destruction"):
            return destruction_list.status in [
                ListStatus.ready_to_review,
                ListStatus.new,
            ]

        return False


class CanDownloadReport(permissions.BasePermission):
    message = _("You cannot download the destruction report.")

    def has_permission(self, request, view):
        return request.user.has_perm("accounts.can_start_destruction")

    def has_object_permission(self, request, view, destruction_list):
        return destruction_list.status == ListStatus.deleted

File: frontend/src/lib/auth/permissions.ts
import {
  STATUSES_ELIGIBLE_FOR_EDIT,
  STATUS_ELIGIBLE_TO_REASSIGN_LIST,
} from "../../pages/constants";
import { User } from "../api/auth";
import { DestructionList } from "../api/destructionLists";

export type PermissionCheck = (user: User) => boolean;

export type DestructionListPermissionCheck = (
  user: User,
  destructionList: DestructionList,
) => boolean;

export const canConfigureApplication: PermissionCheck = (user) => {
  return user.role.canConfigureApplication;
};

/**
 * Returns whether `user` is allowed to create a new destruction list.
 * @param user
 */
export const canStartDestructionList: PermissionCheck = (user) =>
  user.role.canStartDestruction;

/**
 * Returns whether `user` is allowed to delete `destructionList`.
 * @param user
 * @param destructionList
 */
export const canDeleteDestructionList: DestructionListPermissionCheck = (
  user,
  destructionList,
) => {
  if (!user.role.canStartDestruction) {
    return false;
  }
  return destructionList.status === "new";
};

/**
 * Returns whether `user` is allowed to mark `destructionList` as ready to review.
 * @param user
 * @param destructionList
 */
export const canMarkAsReadyToReview: DestructionListPermissionCheck = (
  user,
  destructionList,
) =>
  canStartDestructionList(user) &&
  (destructionList.status === "new" ||
    destructionList.status === "changes_requested");

/**
 * Returns whether `user` is allowed to review `destructionList`.
 * @param user
 * @param destructionList
 */
export const canReviewDestructionList: DestructionListPermissionCheck = (
  user,
  destructionList,
) => {
  if (user.pk !== destructionList.assignee.pk) {
    return false;
  }

  if (
    user.role.canReviewDestruction &&
    destructionList.status === "ready_to_review"
  ) {
    return true;
  }
  if (
    user.role.canReviewFinalList &&
    destructionList.status === "ready_for_archivist"
  ) {
    return true;
  }

  return false;
};

/**
 * Returns whether `user` is allowed to co-review `destructionList`.
 * @param user
 * @param destructionList
 */
export const canCoReviewDestructionList: DestructionListPermissionCheck = (
  user,
  destructionList,
) => {
  if (!user.role.canCoReviewDestruction) {
    return false;
  }

  if (destructionList.status !== "ready_to_review") {
    return false;
  }

  const listAssignees = destructionList.assignees
    .filter((a) => a.role === "co_reviewer")
    .map((a) => a.user.pk);
  if (listAssignees.includes(user.pk)) {
    return true;
  }

  return false;
};

/**
 * Returns whether `user` is allowed to update `destructionList`.
 * @param user
 * @param destructionList
 */
export const canUpdateDestructionList: DestructionListPermissionCheck = (
  user,
  destructionList,
) => {
  if (!user.role.canStartDestruction) {
    return false;
  }

  if (
    destructionList.status === "ready_to_delete" &&
    destructionList.plannedDestructionDate &&
    destructionList.processingStatus === "new"
  ) {
    return false;
  }

  if (!STATUSES_ELIGIBLE_FOR_EDIT.includes(destructionList.status)) {
    return false;
  }

  return true;
};

export const canViewDestructionList: DestructionListPermissionCheck = (
  user,
  destructionList,
) =>
  canStartDestructionList(user) ||
  canReviewDestructionList(user, destructionList) ||
  canCoReviewDestructionList(user, destructionList);

export const canMarkListAsFinal: DestructionListPermissionCheck = (
  user,
  destructionList,
) =>
  canStartDestructionList(user) &&
  destructionList.status === "internally_reviewed";

export const canTriggerDestruction: DestructionListPermissionCheck = (
  user,
  destructionList,
) =>
  canStartDestructionList(user) && destructionList.status === "ready_to_delete";

export const canReassignDestructionList: DestructionListPermissionCheck = (
  user,
  destructionList,
) => {
  return (
    canStartDestructionList(user) &&
    STATUS_ELIGIBLE_TO_REASSIGN_LIST.includes(destructionList.status)
  );
};

/**
 * @param user
 * @param destructionList
 */
export const canUpdateCoReviewers: DestructionListPermissionCheck = (
  user,
  destructionList,
) => {
  if (user.role.canStartDestruction) {
    return (
      destructionList.status === "new" ||
      destructionList.status === "ready_to_review"
    );
  }

  if (user.role.canReviewDestruction) {
    return destructionList.status === "ready_to_review";
  }

  return false;
};

export const canDownloadReport: DestructionListPermissionCheck = (
  user,
  destructionList,
) => {
  if (!user.role.canStartDestruction) {
    return false;
  }
  return (
    destructionList.status === "deleted" &&
    destructionList.processingStatus === "succeeded"
  );
};

export const canRenameDestructionList: DestructionListPermissionCheck = (
  user,
  destructionList,
) => canStartDestructionList(user) && destructionList.status === "new";
